### PR TITLE
Aml 3692 aorn lookup

### PR DESCRIPTION
**Pull request by Winsto**

### CODE REVIEW

File: src/PensionsRegulatorApi/Application/Queries/GetValidatedOrganisationsHandler.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace PensionsRegulatorApi.Application.Queries
+{
+    public class GetValidatedOrganisationsHandler : RequestHandler<GetValidatedOrganisations, IEnumerable<Organisation>>
+    {
+        private readonly IOrganisationRepository _repository;
+
+        public GetValidatedOrganisationsHandler(IOrganisationRepository repository)
+        {
+            _repository = repository ?? throw new ArgumentNullException(nameof(repository));
```
Comment (MichaelYoung1981) [inline src/PensionsRegulatorApi/Application/Queries/GetValidatedOrganisationsHandler.cs]: Pet hate...the null check will never ever be hit as the DI container won't be able to instantiate the class.

File: src/PensionsRegulatorApi/Controllers/PensionsRegulatorController.cs [hunk 2]
```diff
@@ -48,5 +48,33 @@ public async Task<ActionResult<IEnumerable<Organisation>>> Get([FromQuery] strin
...
+        /// <response code="200">Success</response>
+        /// <response code="401">The client is not authorized to access this endpoint</response>
+        /// <response code="500">Internal server error</response>
+        [ProducesResponseType(200)]
+        [ProducesResponseType(401)]
+        [ProducesResponseType(500)]
+        [HttpGet("organisations/{aorn}")]
+        public async Task<ActionResult<IEnumerable<Organisation>>> Get(
+            [FromQuery] string payeRef,
+            [FromRoute] string aorn)
```
Comment (MichaelYoung1981) [inline src/PensionsRegulatorApi/Controllers/PensionsRegulatorController.cs]: I'm being very picky...can aorn go before payeRef?

File: src/PensionsRegulatorApi/Data/SqlOrganisationRepository.cs [hunk 3]
```diff
@@ -60,5 +60,62 @@ public IEnumerable<Organisation> GetOrganisationsForPAYEReference(string payeRef
 
             return retrievedOrganisations.AsReadOnly();
         }
+
+        public IEnumerable<Organisation> GetOrganisationsForPAYEReferenceAndAORN(string payeReference, string aorn)
```
Comment (MichaelYoung1981) [inline src/PensionsRegulatorApi/Data/SqlOrganisationRepository.cs]: Can you have a look at refactoring the method?  It feels too long.

File: src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReferenceAndAORN.sql [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+Organisation.OrganisationName as Name,
+Organisation.Status as Status,
+Organisation.TPR_Unique_Id as UniqueIdentity,
+OrganisationAddress.AddressLine1 as AddressLine1,
+OrganisationAddress.AddressLine2 as AddressLine2,
+OrganisationAddress.AddressLine3 as AddressLine3,
+OrganisationAddress.AddressLine4 as AddressLine4,
+OrganisationAddress.AddressLine5 as AddressLine5,
+OrganisationAddress.PostCode as PostCode
+FROM
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReferenceAndAORN.sql]: As, I presume, this stored proc and GetOrganisationsByPAYEReference are doing the same joins and using the same columns would it be worth creating a view and simplifying the stored procs?

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+{
+    [TestFixture]
+    [ExcludeFromCodeCoverage]
+    public class Given_Repository_Returns_Results
+    {
+        private IOrganisationRepository _mockRepository;
+        private IRequestHandler<global::PensionsRegulatorApi.Application.Queries.GetValidatedOrganisations, IEnumerable<Organisation>> _sut;
+        private IEnumerable<Organisation> _repositoryResults;
+
+        public Given_Repository_Returns_Results()
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs]: Can we make the test names less implementation specific?  We need to test the expected behaviour not how we're doing it.

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                            .Next(1, 15));
+        }
+
+        [SetUp]
+        public void Given()
+        {
+            _mockRepository = Substitute.For<IOrganisationRepository>();
+
+            _mockRepository
+                .GetOrganisationsForPAYEReferenceAndAORN(Arg.Any<string>(), Arg.Any<string>())
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs]: Any should be show with sh1t for setups, can we use fixed values?

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Throws_An_Exception.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+using NUnit.Framework;
+using PensionsRegulatorApi.Application.Queries;
+using PensionsRegulatorApi.Data;
+using PensionsRegulatorApi.Domain;
+
+namespace SFA.DAS.PensionsRegulatorApi.UnitTests.Application.Queries.GetValidatedOrganisations
+{
+    [TestFixture]
+    [ExcludeFromCodeCoverage]
+    public class Given_Repository_Throws_An_Exception
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Throws_An_Exception.cs]: Same as previous class.

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Controllers/PayeAndAorn/Given_No_Results_Are_Returned.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+using NUnit.Framework;
+using PensionsRegulatorApi.Application.Queries;
+using PensionsRegulatorApi.Controllers;
+using PensionsRegulatorApi.Domain;
+
+namespace SFA.DAS.PensionsRegulatorApi.UnitTests.Controllers.PayeAndAorn
+{
+    [TestFixture]
+    [ExcludeFromCodeCoverage]
+    public class Given_No_Results_Are_Returned
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.PensionsRegulatorApi.UnitTests/Controllers/PayeAndAorn/Given_No_Results_Are_Returned.cs]: Naming same as previous comment, less technical names would be good.

File: src/PensionsRegulatorApi/Controllers/PensionsRegulatorController.cs (hunk 2 above)
Comment (tom-gough) [inline src/PensionsRegulatorApi/Controllers/PensionsRegulatorController.cs]: Not a fan of one being route param and one being query param. Makes it look like there is a logic/business reason other than the / decoding issue. Assume this was something you guys have discussed and agreed in the office though in which case I will pipe down.

File: src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReference.sql [hunk 9]
```diff
@@ -1,5 +1,5 @@
 ﻿CREATE PROCEDURE [dbo].GetOrganisationsByPAYEReference
-	@PAYESchemeReference nvarchar(50)
+	@PAYESchemeReference varchar(25)
```
Comment (tom-gough) [inline src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReference.sql]: will it ever be this long?

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        [Test]
+        public async Task Then_Equivalent_Results_Are_Returned()
+        {
+            var organisations =
+                await
+                    _sut
+                        .Handle(
+                            new global::PensionsRegulatorApi.Application.Queries.GetValidatedOrganisations(
+                                "anynonemptytext",
+                                "anynonemptytext"),
```
Comment (tom-gough) [inline src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs]: can just be "" or string.Empty I think given your setup

File: src/PensionsRegulatorApi/Controllers/PensionsRegulatorController.cs (hunk 2 above)
Comment (Winsto) [reply src/PensionsRegulatorApi/Controllers/PensionsRegulatorController.cs]: @tom-gough  this is as a result of the discussion we had around the fact that if both are query params then we end up with one method and switching based on null/empty checks on the aorn param.

This 'hybrid' route/query approach was so we could have two distinct methods.

We can comment to explain though.

File: src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReference.sql (hunk 9 above)
Comment (Winsto) [reply src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReference.sql]: I am taking the data types from the DB design doc, so just doing as I'm told here :)

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs (hunk 10 above)
Comment (Winsto) [reply src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs]: Following some of @MichaelYoung1981 's comments these query tests have been completely re-written.

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs (hunk 6 above)
Comment (Winsto) [reply src/SFA.DAS.PensionsRegulatorApi.UnitTests/Application/Queries/GetValidatedOrganisations/Given_Repository_Returns_Results.cs]: I have now re-written the query tests in a different style; you may not like this new style, so I may have to go back to GWT and rename :)

Whilst doing this I decided to rename the queries to be more representative as well (specifically 'validate' was no longer being used in the api, so having queries with that name didn't make sense).

File: src/PensionsRegulatorApi/Data/SqlOrganisationRepository.cs (hunk 3 above)
Comment (Winsto) [reply src/PensionsRegulatorApi/Data/SqlOrganisationRepository.cs]: Done some work on this and reduced a fair amount of repetition.

File: src/SFA.DAS.PensionsRegulatorApi.UnitTests/Controllers/PayeAndAorn/Given_No_Results_Are_Returned.cs (hunk 8 above)
Comment (Winsto) [reply src/SFA.DAS.PensionsRegulatorApi.UnitTests/Controllers/PayeAndAorn/Given_No_Results_Are_Returned.cs]: Done a rewrite of the controller tests now, see what you think.

File: src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReferenceAndAORN.sql (hunk 4 above)
Comment (Winsto) [reply src/SFA.DAS.PensionsRegulator.Database/StoredProcedures/GetOrganisationsByPAYEReferenceAndAORN.sql]: View created an sprocs modified to read from view.

**Comment by Winsto** [comment]
Now rendered redundant by [combined branch](https://github.com/SkillsFundingAgency/das-pensionsregulator/pull/21).

**Pull request closed without merging**